server/download.go

```diff
@@ -44,17 +44,19 @@ type blobDownload struct {
 
 	context.CancelFunc
 
-	done       bool
+	done       chan struct{}
 	err        error
 	references atomic.Int32
 }
 
 type blobDownloadPart struct {
-	N           int
-	Offset      int64
-	Size        int64
-	Completed   int64
-	lastUpdated time.Time
+	N         int
+	Offset    int64
+	Size      int64
+	Completed atomic.Int64
```

Comment:
there shouldn't be any concurrent operations on this since this tracks the completion state of one part. the full download uses an atomic.Int64 because multiple concurrent parts will update to it but a part should update its own state

Reply:
Run ollama serve with `-race` and perform a pull. You'll find there is. See everywhere below where it is accessed for writes while also for reads while updated progress.